micro-rdk/src/common/data_store.rs

```diff
@@ -0,0 +1,480 @@
+/// reserved in DATA_STORE. The memory is segmented based according to the DataCollectors expected
+/// (identified by collector keys) and each segment view is treated as a separate ring buffer of SensorData
+/// messages. Currently, an equal amount of space is alloted to each collector, which will affect
+/// the maximum allowed size of a single message (computed as the length of DATA_STORE divided by
+/// the number of collector keys). It should be treated as a global struct that should only be initialized once
+/// and is not thread-safe (all interactions should be blocking).
+pub struct StaticMemoryDataStore {
+    buffers: Vec<LocalRb<u8, &'static mut [MaybeUninit<u8>]>>,
+    collector_keys: Vec<ResourceMethodKey>,
+}
+
+impl StaticMemoryDataStore {
+    pub fn new(collector_keys: Vec<ResourceMethodKey>) -> Result<Self, DataStoreError> {
+        if !DATA_STORE_INITIALIZED.fetch_or(true, Ordering::SeqCst) {
+            let len_per_buffer = unsafe { DATA_STORE.len() } / collector_keys.len();
+            let mut buffers = Vec::new();
+            for i in 0..(collector_keys.len()) {
+                let start_idx = i * len_per_buffer;
+                let end_idx = (i + 1) * len_per_buffer;
+                unsafe {
+                    buffers.push(LocalRb::from_raw_parts(
+                        &mut DATA_STORE[start_idx..end_idx],
+                        0,
+                        0,
+                    ));
+                }
+            }
+            return Ok(Self {
+                buffers,
+                collector_keys,
+            });
+        }
+        Err(DataStoreError::DataStoreInitialized)
+    }
+
+    fn get_index_for_collector(
+        &self,
+        collector_key: &ResourceMethodKey,
+    ) -> Result<usize, DataStoreError> {
+        self.collector_keys
+            .iter()
+            .position(|key| key == collector_key)
+            .ok_or(DataStoreError::UnknownCollectorKey(collector_key.clone()))
+    }
+}
+
+impl DataStore for StaticMemoryDataStore {
+    fn store_message(
+        &mut self,
+        collector_key: ResourceMethodKey,
+        message: SensorData,
+    ) -> Result<(), DataStoreError> {
+        let buffer_index = self.get_index_for_collector(&collector_key)?;
+        let buffer = &mut self.buffers[buffer_index];
+        let encode_len = message.encoded_len();
+        if encode_len + 5 > buffer.vacant_len() {
+            return Err(DataStoreError::DataBufferFull(collector_key, message));
+        }
+        buffer
+            .push(0)
+            .map_err(|_| DataStoreError::DataWriteFailure)?;
+        let len_bytes = (encode_len as u32).to_le_bytes();
+        buffer.push_slice(&len_bytes);
+
+        unsafe {
+            let mut prod = Producer::new(buffer as &LocalRb<u8, &'static mut [MaybeUninit<u8>]>);
+            let (left, right) = prod.free_space_as_slices();
+            let mut chained = BufMut::chain_mut(left, right);
+            message.encode(&mut chained)?;
```

Comment:
While looking at the Prost Message regarding the above, I noticed that it has an `encode_length_delimited` and `decode_length_delimited` pair. Maybe that could be used instead of manually encoding the length above?

Reply:
So the reason that I didn't use those methods is because they encode the length as a `u64` which felt unnecessarily space inefficient. If you feel the trade-off for the simplicity is worth it though, I can change

Reply:
I believe it encodes as varint which could actually do better than u32 for small messages:
- https://docs.rs/prost/latest/src/prost/message.rs.html#89
- https://protobuf.dev/programming-guides/encoding/#varints
You would need to use `encoded_len_varint` too, I think, to know how many total bytes were required.